Sources/AsyncHTTPClient/ConnectionPool/HTTPRequestStateMachine.swift

```diff
@@ -0,0 +1,467 @@
+        }
+    }
+
+    mutating func start() -> Action {
+        guard case .initialized = self.state else {
+            preconditionFailure("Invalid state")
+        }
+
+        self.state = .running(.verifyRequest, .initialized)
+        return .verifyRequest
+    }
+
+    mutating func requestVerified(_ head: HTTPRequestHead) -> Action {
+        guard case .running(.verifyRequest, .initialized) = self.state else {
+            preconditionFailure("Invalid state")
+        }
+
+        guard self.isChannelWritable else {
+            preconditionFailure("Unimplemented. Wait with starting the request here!")
+        }
+
+        if let value = head.headers.first(name: "content-length"), let length = Int(value), length > 0 {
+            self.state = .running(.streaming(expectedBodyLength: length, sentBodyBytes: 0, producer: .producing), .initialized)
+            return .sendRequestHead(head, startBody: true, startReadTimeoutTimer: nil)
+        } else if head.headers.contains(name: "transfer-encoding") {
+            self.state = .running(.streaming(expectedBodyLength: nil, sentBodyBytes: 0, producer: .producing), .initialized)
+            return .sendRequestHead(head, startBody: true, startReadTimeoutTimer: nil)
+        } else {
+            self.state = .running(.endSent, .initialized)
+            return .sendRequestHead(head, startBody: false, startReadTimeoutTimer: self.idleReadTimeout)
+        }
+    }
+
+    mutating func requestVerificationFailed(_ error: Error) -> Action {
+        guard case .running(.verifyRequest, .initialized) = self.state else {
+            preconditionFailure("Invalid state")
+        }
+
+        self.state = .failed(error)
+        return .failRequest(error, closeStream: false)
+    }
+
+    mutating func requestStreamPartReceived(_ part: IOData) -> Action {
+        switch self.state {
+        case .initialized:
+            preconditionFailure("Invalid state: \(self.state)")
+
+        case .running(.verifyRequest, _),
+             .running(.endSent, _):
+            preconditionFailure("Invalid state: \(self.state)")
+
+        case .running(.streaming(let expectedBodyLength, var sentBodyBytes, let producerState), let responseState):
+            // More streamed data is accepted, even though the producer should stop. However
+            // there might be thread syncronisations situations in which the producer might not
+            // be aware that it needs to stop yet.
+
+            if let expected = expectedBodyLength {
+                if sentBodyBytes + part.readableBytes > expected {
+                    let error = HTTPClientError.bodyLengthMismatch
+
+                    switch responseState {
+                    case .initialized, .receivingBody:
+                        self.state = .failed(error)
+                    case .endReceived:
+                        #warning("TODO: This needs to be fixed. @Cory: What does this mean here?")
```

Comment:
As discussed, this represents an error condition, and we should fail the request.